core/src/main/scala/org/apache/spark/util/collection/PartitionedSerializedPairBuffer.scala

```diff
@@ -246,9 +247,13 @@ private[spark] class SerializedSortDataFormat extends SortDataFormat[Int, IntBuf
 }
 
 private[spark] object PartitionedSerializedPairBuffer {
-  val KEY_START = 0
-  val VAL_START = 1
-  val VAL_END = 2
+  val KEY_START = 0 // keyStart, a long, gets split across two ints
+  val KEY_VAL_LEN = 2
   val PARTITION = 3
-  val RECORD_SIZE = Seq(KEY_START, VAL_START, VAL_END, PARTITION).size // num ints of metadata
+  val RECORD_SIZE = PARTITION + 1 // num ints of metadata
+
+  def keyStartPos(metaBuffer: IntBuffer, metaBufferPos: Int): Long = {
+    metaBuffer.get(metaBufferPos + KEY_START) +
```

Comment:
Yeah, this isn't quite right:

```scala
scala> val long: Long = 0x180000000L
long: Long = 6442450944

scala> val upper32: Int = (long >> 32).toInt
upper32: Int = 1

scala> val lower32: Int = long.toInt
lower32: Int = -2147483648

scala> (upper32.toLong << 32) + lower32
res0: Long = 2147483648
```

Instead of adding these numbers together, you should do bitwise OR them and use a mask to handle the treating the lower 32 bits as an unsigned integer rather than having it be converted into a signed long:

```scala
scala> (upper32.toLong << 32) | (lower32 & 0xFFFFFFFFL)
res17: Long = 6442450944
```